File: server/src/models/OrderProduct.ts
import {
  Table,
  Model,
  Column,
  IsUUID,
  PrimaryKey,
  Default,
  DataType,
  BelongsTo,
  ForeignKey,
  AllowNull,
  Is,
  BeforeCreate,
} from "sequelize-typescript";
import { Field, ObjectType } from "type-graphql";
import { Product } from "./Product";
import { Order } from "./Order";
import { Transaction } from "sequelize/types";

@ObjectType()
@Table({
  timestamps: true,
  underscored: true,
})
export class OrderProduct extends Model {
  @Field()
  @IsUUID(4)
  @PrimaryKey
  @Default(DataType.UUIDV4)
  @Column
  id: string;

  @IsUUID(4)
  @ForeignKey(() => Order)
  @AllowNull(false)
  @Column
  orderId: string;

  @BelongsTo(() => Order)
  order: Order;

  @IsUUID(4)
  @ForeignKey(() => Product)
  @AllowNull(false)
  @Column
  productId: string;

  @BelongsTo(() => Product)
  product: Product;

  @Field()
  @Default(1)
  @Is("Positive", (value) => {
    if (value <= 0) {
      throw new Error(`"${value}" must be greater than 0.`);
    }
  })
  @AllowNull(false)
  @Column
  count: number;

  @Field()
  @Is("Positive", (value) => {
    if (value < 0) {
      throw new Error(`"${value}" must be positive.`);
    }
  })
  @Column
  price: number;

  @Field()
  createdAt: Date;

  @Field()
  updatedAt: Date;

  @BeforeCreate
  static async canOrderQuantity(
    instance: OrderProduct,
    { transaction }: { transaction: Transaction }
  ) {
    const product = await Product.findOne({
      where: { id: instance.productId },
      transaction,
    });

    if (!product?.isActive) {
      throw new Error("This product is inactive.");
    }

    if (product?.isMadeToOrder && product.quantity < instance.count) {
      throw new Error(
        "You cannot order more than what is available unless it can be made to order."
      );
    }
  }
}

File: go-server/internal/models/order_product.go
package models

import (
	"errors"
	"fmt"

	"gorm.io/gorm"
)

func init() {
	registerModel(&OrderProduct{})
}

// OrderProduct is a line item on a placed Order: a snapshot of one product,
// its price at the time of purchase, and the quantity ordered. Only
// count/price/createdAt/updatedAt carry @Field() on the Node model —
// orderId/productId are DB-only, "product" is a FieldResolver
// (OrderedProductResolver.Product), so it's forced to a resolver in
// gqlgen.yml despite the matching struct field below.
type OrderProduct struct {
	Base

	OrderID string `gorm:"not null;index" json:"-"`
	Order   *Order `gorm:"foreignKey:OrderID;constraint:OnDelete:CASCADE" json:"-"`

	ProductID string   `gorm:"not null;index" json:"-"`
	Product   *Product `gorm:"foreignKey:ProductID;constraint:OnDelete:CASCADE" json:"-"`

	Count int `gorm:"not null;default:1" json:"count"`
	Price int `json:"price"`
}

// BeforeSave mirrors the `@Default(1)` + `@Is("Positive", ...)` on count
// (see UserProduct.BeforeSave for why the default is applied here too) and
// the plain `@Is("Positive", ...)` on price (no default — 0 is a valid
// price, so unlike count there's nothing to backfill).
func (op *OrderProduct) BeforeSave(tx *gorm.DB) error {
	if op.Count == 0 {
		op.Count = 1
	}
	if op.Count <= 0 {
		return fmt.Errorf(`"%d" must be greater than 0.`, op.Count)
	}
	if op.Price < 0 {
		return fmt.Errorf(`"%d" must be positive.`, op.Price)
	}
	return nil
}

// BeforeCreate mirrors OrderProduct.canOrderQuantity: the product must be
// active, and can't be ordered past what's on hand unless it's
// made-to-order.
func (op *OrderProduct) BeforeCreate(tx *gorm.DB) error {
	op.EnsureID()

	var product Product
	if err := tx.Where("id = ?", op.ProductID).First(&product).Error; err != nil {
		return err
	}
	if !product.IsActive {
		return errors.New("This product is inactive.")
	}
	if !product.IsMadeToOrder && product.Quantity < op.Count {
		return errors.New("You cannot order more than what is available unless it can be made to order.")
	}
	return nil
}

// AfterCreate mirrors OrderProduct.reduceAvailability: every cart line for
// this product across all users shrinks (or disappears) by the amount just
// ordered, since that much is no longer available.
func (op *OrderProduct) AfterCreate(tx *gorm.DB) error {
	var userProducts []UserProduct
	if err := tx.Where("product_id = ?", op.ProductID).Find(&userProducts).Error; err != nil {
		return err
	}

	for i := range userProducts {
		up := &userProducts[i]
		if up.Count-op.Count > 0 {
			up.Count -= op.Count
			if err := tx.Save(up).Error; err != nil {
				return err
			}
		} else if err := tx.Delete(up).Error; err != nil {
			return err
		}
	}
	return nil
}
